compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2713,12 +2713,13 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                         Some(t) if t.hir_owner == parent_id => t,
                         _ => self.tcx.typeck(parent_id),
                     };
-                    let ty = typeck_results.expr_ty_adjusted(expr);
-                    let span = expr.peel_blocks().span;
+                    let expr = expr.peel_blocks();
```

Comment:
peeling first then looking for the expr lets us recover the type in `ruby_style_closure`, but it's still insufficient for the UI test I added, but oh well. It's probably just not being recorded for some other typeck reason.